A47-xds-federation.md

```diff
@@ -0,0 +1,684 @@
+    // "xdstp://<authority_name>/envoy.config.listener.v3.Listener/%s".
+    "client_listener_resource_name_template": <string>,
+
+    // Ordered list of xDS servers to contact for this authority.
+    // Format is exactly the same as the top level "xds_servers" field.
+    //
+    // If the same server is listed in multiple authorities, the
+    // entries will be de-duped (i.e., resources for both authorities
+    // will be fetched on the same ADS stream).
+    //
+    // If not specified, the top-level server list is used.
+    "xds_servers": [ ... ]
+  }
+}
+```
+
+The new `client_default_listener_resource_name_template` field will have
+the following format:
+
+```
+// A template for the name of the Listener resource to subscribe to
+// for a gRPC client channel.  Used only when the channel is created
+// with an "xds:" URI with no authority.
+//
+// If starts with "xdstp:", will be interpreted as a new-style name,
+// in which case the authority of the URI will be used to select the
+// relevant configuration in the "authorities" map.
+//
+// The token "%s", if present in this string, will be replaced with
+// the service authority (i.e., the path part of the target URI
+// used to create the gRPC channel).  If the template starts with
+// "xdstp:", the replaced string will be percent-encoded.
+//
+// Defaults to "%s".
+"client_default_listener_resource_name_template": <string>,
+```
+
+There will be two changes to the existing
+`server_listener_resource_name_template` described in [gRFC A36][A36]:
+- If the template starts with `xdstp:`, it will be interpreted as a
+  new-style name, in which case the authority of the URI will be used to
+  select the relevant configuration in the `authorities` map.
+- When replacing the `%s` token, if the template starts with `xdstp:`,
+  the replacement string will be percent-encoded in the resulting resource
+  name.
+
+The resulting field will have the following format:
+
+```
+// A template for the name of the Listener resource to subscribe to
+// for a gRPC server.
+//
+// If starts with "xdstp:", will be interpreted as a new-style name,
+// in which case the authority of the URI will be used to select the
+// relevant configuration in the "authorities" map.
+//
+// The token "%s", if present in this string, will be replaced with
+// the IP and port on which the server is listening.  If the template
+// starts with "xdstp:", the replaced string will be percent-encoded.
+//
+// There is no default; if unset, xDS-based server creation fails.
+//
+// Example:
+// "xdstp://xds.authority.com/envoy.config.listener.v3.Listener/grpc/server/%s"
+"server_listener_resource_name_template": <string>,
+```
+
+#### Example: Bootstrap Config Contains No New Fields
+
+If the bootstrap config does not contain any of the new fields described
+above, then here's how the resource names will be determined.
+
+If a gRPC client channel is created for `xds:server.example.com`:
```

Comment:
Should we use `xds:///server.example.com`?  In java `xds:server.example.com` is not a URI String

Reply:
As per RFC-3986, `xds:server.example.com` is a valid URI with scheme `xds` and path `server.example.com`.

I'm aware that Java doesn't currently handle this format.  @ejona86 and I have discussed this in the past, and he agrees that it needs to be fixed, but I don't think that work has been done yet.